ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs

```diff
@@ -268,3 +307,195 @@ mkProtocolRealTPraos genesis CoreNode { cnDelegateKey, cnVRF, cnKES, cnOCert } =
+      , stcNodeRestarts = nodeRestarts
+      }
+
+shrinkShelleyTestConfig :: ShelleyTestConfig -> [ShelleyTestConfig]
+shrinkShelleyTestConfig _ = [] -- TODO
+
+{-------------------------------------------------------------------------------
+  Node restarts for operational certificates
+-------------------------------------------------------------------------------}
+
+-- | Generate a node restart schedule which cycles nodes in order to replace
+-- their operational certificates after the maximum number of KES periods.
+--
+ocertNodeRestarts
+  :: NodeJoinPlan
+  -> NumSlots         -- ^ Duration of the test. No replacements will be scheduled
+                      --   after this number of slots.
+  -> ShelleyGenesis c -- ^ Shelley genesis configuration. Needed to extract the
+                      --   slots per KES period and the maximum number of KES
+                      --   evolutions.
+  -> [CoreNode c]     -- ^ Core nodes, including all of their keys.
+  -> (Word64, Word64) -- ^ Permitted (eagerness, laziness) in replacing operational
+                      --   certificates - that is, nodes may replace them up to
+                      --   `eagerness` KES periods early, and up to `laziness`
+                      --   KES periods late.
+  -> Gen NodeRestarts
+ocertNodeRestarts
+  (NodeJoinPlan _m)
+  (NumSlots t)
+  ShelleyGenesis {sgSlotsPerKESPeriod, sgMaxKESEvolutions}
+  coreNodes
+  (eagerness, laziness)
+    | t < ocertRegenCadence = pure noRestarts
+    | otherwise =
+      fmap NodeRestarts $ do
+        slotCNPairs <- fmap join $ forM (zip coreNodes [0..]) $ \(cn, coreNodeIdx) -> do
+          offsets <- Gen.infiniteListOf $ Gen.choose (- eagerness, laziness)
+          let SL.KESPeriod firstKESPeriod = SL.ocertKESPeriod $ cnOCert cn
+              periods =
+                unfoldr
+                  ( \(startPeriod, ix) -> let
+                      np = fromIntegral startPeriod + sgMaxKESEvolutions + offsets !! ix
+                      in Just $
+                    ( fromIntegral np
+                    , (fromIntegral np, ix + 1)
+                    )
+                  )
+                  (firstKESPeriod, 0)
+          pure
+            [ (SlotNo slot, CoreNodeId coreNodeIdx)
+              | slot <- takeWhile (< t) ((kesPeriodStartSlots !!) <$> periods)
+            ]
+        pure
+          . Map.map (Map.fromList . fmap (,NodeRekey))
+          . Map.fromListWith (<>)
+          $ fmap (: []) <$> slotCNPairs
+    where
+      -- KES period i begins in slot kesPeriodStartSlots !! i
+      kesPeriodStartSlots = [i * sgSlotsPerKESPeriod | i <- [0 ..]]
+      ocertRegenCadence = sgSlotsPerKESPeriod * sgMaxKESEvolutions
+
+ocertRekeying
+  :: forall c m. (TPraosCrypto c)
+  => Seed
+  -> ShelleyGenesis c
+  -> [CoreNode c]
+  -> Rekeying m (ShelleyBlock c)
+ocertRekeying
+  initSeed
+  ShelleyGenesis {sgSlotsPerKESPeriod}
+  coreNodes
+  = Rekeying
+  { rekeyOracle = \_ sn -> Just sn
```

Comment:
@nc6 Would you revisit the reason you mentioned for this being a valid `rekeyOracle` and at it as a comment here?

Reply:
Err, I now see `tniCrucialTxs = []` below. So that's probably why the oracle doesn't matter. This rekey doesn't require posting anything on the chain, so timeliness/promptness isn't a concern like it was for Byron.